### PR TITLE
Added sidebar toggler

### LINKED ISSUE #196: ui: add the sidebar toggle
It would be great if users will have an ability to hide/show the Sidebar.

1) Add the fixed "Collapse" button to the left bottom corner of the Content zone
2) Click on this button should hide the Sidebar complitely
3) When the Sidebar is hidden, the icon changes to the "icon right". Click to this button will show the Sidebar back.
4) The sidebar visibility state should be saved at the LocalStorage 
5) Add the "CMD+." shortcut to the Sidebar toggling. Use the [CodeX Shortcuts](https://github.com/codex-team/codex.shortcuts) lib for that, for example.

image

### PR DESCRIPTION
**Pull request by Y-k-Y**
Resolves #196

* The shortcut is "CMD + SHIFT + S" since "+" character is not supported at shortcut package

### CODE REVIEW

**Comment by neSpecc** [comment]
> The shortcut is "CMD + SHIFT + S" since "+" character is not supported at shortcut package

The shortcut should be "CMD + ." since it is the common shortcuts for hiding panels (Figma, Sketch, etc). The plus is not the key, just a joiner for combination of the CMD and the Dot

File: src/frontend/js/modules/sidebar.js [hunk 1]
```diff
@@ -45,14 +50,21 @@ export default class Sidebar {
...
     };
     this.sidebarStorage = new Storage(LOCAL_STORAGE_KEY);
     const storedState = this.sidebarStorage.get();
 
     this.sectionsState = storedState ? JSON.parse(storedState) : {};
+
+    this.sidebarVisibilityStorage = new Storage(SIDEBAR_VISIBILITY_KEY);
+    const storedVisibility = this.sidebarVisibilityStorage.get();
+
+    this.visibility = storedVisibility !== 'false';
```
Comment (neSpecc) [inline src/frontend/js/modules/sidebar.js:67]: Please, give more clear property name. For example, `isVisible`

File: src/frontend/styles/components/sidebar.pcss [hunk 2]
```diff
@@ -1,6 +1,34 @@
...
+
+  &--animated {
+    .docs-sidebar__content {
+      transition: transform 200ms ease-in-out;
+    }
+
+    .docs-sidebar__slider {
+      transition: left 200ms ease-in-out;
+
+      svg {
```
Comment (neSpecc) [inline src/frontend/styles/components/sidebar.pcss:12]: lets remove transition from the SVG sice the whole button is moving at this moment, and the svg rotation animation is looks not so well.

File: src/frontend/styles/components/sidebar.pcss [hunk 3]
```diff
@@ -1,6 +1,34 @@
 .docs-sidebar {
   width: 100vw;
-    
+
+  &--animated {
+    .docs-sidebar__content {
+      transition: transform 200ms ease-in-out;
```
Comment (neSpecc) [inline src/frontend/styles/components/sidebar.pcss:6]: Please, add the "will-change: transform" to the `.docs-sidebar__content`'s default rules set to optimize the repaint

File: src/frontend/styles/components/sidebar.pcss [hunk 4]
```diff
@@ -1,6 +1,34 @@
...
+    }
+  }
+
+  &--collapsed {
+    .docs-sidebar__content {
+      transform: translateX(-100%);
+    }
+
+    .docs-sidebar__slider {
+      left: 20px;
```
Comment (neSpecc) [inline src/frontend/styles/components/sidebar.pcss:24]: it's better to use `translate` instead of `left` since it does not fire the reflow.

**Review by neSpecc — CHANGES_REQUESTED** [review]

**Review by neSpecc — CHANGES_REQUESTED** [review]
The Sidebar should be always visible on mobile layout independent on the `collapsed` state

image

**Comment by Y-k-Y** [comment]
> > The shortcut is "CMD + SHIFT + S" since "+" character is not supported at shortcut package
> 
> The shortcut should be "CMD + ." since it is the common shortcuts for hiding panels (Figma, Sketch, etc). The plus is not the key, just a joiner for combination of the CMD and the Dot

Changed shortcut as "Ctrl + ."

File: src/frontend/js/modules/sidebar.js [hunk 5]
```diff
@@ -168,12 +183,48 @@ export default class Sidebar {
...
+   *
+   * @returns {void}
+   */
+  initSidebar() {
+    if (!this.visibility) {
+      this.nodes.slider.classList.add(Sidebar.CSS.sidebarSliderHidden);
+      this.nodes.sidebar.classList.add(Sidebar.CSS.sidebarCollapsed);
+    }
+
+    // prevent sidebar animation on page load
```
[the remark this answers is not in the document]
Comment (Y-k-Y) [reply src/frontend/js/modules/sidebar.js]: Added description

File: src/frontend/js/modules/sidebar.js [hunk 6]
```diff
@@ -168,12 +183,48 @@ export default class Sidebar {
...
   }
 
+  /**
+   * Initializes sidebar
+   *
+   * @returns {void}
+   */
+  initSidebar() {
+    if (!this.visibility) {
+      this.nodes.slider.classList.add(Sidebar.CSS.sidebarSliderHidden);
```
[the remark this answers is not in the document]
Comment (Y-k-Y) [reply src/frontend/js/modules/sidebar.js]: Removed

File: src/frontend/js/modules/sidebar.js (hunk 1 above)
Comment (Y-k-Y) [reply src/frontend/js/modules/sidebar.js]: Fixed

File: src/frontend/styles/components/sidebar.pcss (hunk 2 above)
Comment (Y-k-Y) [reply src/frontend/styles/components/sidebar.pcss]: Fixed

File: src/frontend/styles/components/sidebar.pcss (hunk 3 above)
Comment (Y-k-Y) [reply src/frontend/styles/components/sidebar.pcss:6]: Fixed

File: src/frontend/styles/components/sidebar.pcss (hunk 4 above)
Comment (Y-k-Y) [reply src/frontend/styles/components/sidebar.pcss]: Fixed

**Comment by Y-k-Y** [comment]
> The Sidebar should be always visible on mobile layout independent on the `collapsed` state
> 
> image

Fixed

File: src/frontend/js/modules/sidebar.js [hunk 7]
```diff
@@ -45,14 +49,22 @@ export default class Sidebar {
...
     this.sidebarStorage = new Storage(LOCAL_STORAGE_KEY);
     const storedState = this.sidebarStorage.get();
 
     this.sectionsState = storedState ? JSON.parse(storedState) : {};
+
+    this.sidebarVisibilityStorage = new Storage(SIDEBAR_VISIBILITY_KEY);
+    const storedVisibility = this.sidebarVisibilityStorage.get();
+
+    this.isVisible = storedVisibility !== 'false';
+    this.keys = {};
```
Comment (neSpecc) [inline src/frontend/js/modules/sidebar.js:67]: Please add jsdocs describing all the new properties you've added to the class

**Comment by slaveeks** [comment]
Why shortcut contains Ctrl, but not CMD? CMD is more comfortable for Mac

File: src/frontend/js/modules/sidebar.js [hunk 8]
```diff
@@ -168,12 +183,63 @@ export default class Sidebar {
...
+     * Since animated class contains transition, hiding will be animated with it
+     * To prevent awkward animation when visibility is set to false, we need to remove animated class
+     */
+    setTimeout(() => {
+      this.nodes.sidebar.classList.add(Sidebar.CSS.sidebarAnimated);
+    }, 200);
+
+    // add event listener to execute keyboard shortcut
+    document.body.addEventListener('keydown', (e) => this.executeSlide(e), false);
+    document.body.addEventListener('keyup', (e) => this.executeSlide(e), false);
```
Comment (neSpecc) [inline src/frontend/js/modules/sidebar.js:207]: You can use CodeX Shortcuts package for that

**Comment by slaveeks** [comment]
Can we make several kinds of shortcut? For Windows and MacOS

**Comment by neSpecc** [comment]
1) Shortcut does not works (tested in Safari, MacOS)
2) Somethimes the animation looks blinks twice

https://user-images.githubusercontent.com/3684889/182409494-7fb74463-7f21-48c5-8266-6db68235a8c7.mp4

File: src/frontend/js/modules/sidebar.js [hunk 9]
```diff
@@ -168,12 +183,63 @@ export default class Sidebar {
...
+   */
+  executeSlide(event) {
+    this.keys[event.key.toLowerCase()] = event.type === 'keydown';
+
+    /**
+     * Execute slide when ctrl + . is pressed
+     */
+    if (this.keys['control'] && this.keys['.']) {
+      this.handleSliderClick();
+    }
```
Comment (robonetphy) [inline src/frontend/js/modules/sidebar.js:222]: no need to maintain `this.key` can do like this
```
if ((e.ctrlKey || e.metaKey) && e.code === 'Period') {
      this.handleSliderClick();
    }
```
but better to use [codex.shortcut](https://github.com/codex-team/codex.shortcuts) module

**Review by robonetphy — CHANGES_REQUESTED** [review]

**Comment by Y-k-Y** [comment]
> 1. Shortcut does not works (tested in Safari, MacOS)
> 2. Somethimes the animation blinks twice
> 
>  Aug-02-2022.18-11-25.mp4

I'll fix this issue tomorrow since i don't have mac now

File: src/frontend/js/modules/sidebar.js (hunk 8 above)
Comment (robonetphy) [reply src/frontend/js/modules/sidebar.js]: @Y-k-Y I have added the support for period you can now use it

**Comment by Y-k-Y** [comment]
> 1. Shortcut does not works (tested in Safari, MacOS)
> 2. Somethimes the animation blinks twice
> 
>  Aug-02-2022.18-11-25.mp4

Fixed

File: src/frontend/js/modules/sidebar.js [hunk 10]
```diff
@@ -45,14 +49,26 @@ export default class Sidebar {
...
+
+    // Initialize localStorage that contains sidebar visibility
+    this.sidebarVisibilityStorage = new Storage(SIDEBAR_VISIBILITY_KEY);
+    // Get current sidebar visibility from storage
+    const storedVisibility = this.sidebarVisibilityStorage.get();
+
+    // Sidebar visibility
+    this.isVisible = storedVisibility !== 'false';
+    // Keys to store currently pressed keys
+    this.keys = {};
```
Comment (robonetphy) [inline src/frontend/js/modules/sidebar.js:71]:
```suggestion
```

**Review by robonetphy — CHANGES_REQUESTED** [review]

File: src/frontend/js/modules/sidebar.js (hunk 10 above)
Comment (robonetphy) [reply src/frontend/js/modules/sidebar.js:71]: pls change this so I can approve it

Comment (Y-k-Y) [reply src/frontend/js/modules/sidebar.js]: just removed!

**Review by robonetphy — APPROVED** [review]

**Review by neSpecc — APPROVED** [review]

**Pull request merged**